plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange.hs

```diff
@@ -0,0 +1,107 @@
+                                                          toNormalizedFilePath',
+                                                          uriToFilePath')
+import           Development.IDE.Core.Actions            (useE)
+import           Development.IDE.Core.PositionMapping    (PositionMapping,
+                                                          fromCurrentPosition,
+                                                          toCurrentRange)
+import           Development.IDE.GHC.Compat              (HieAST (Node), Span,
+                                                          getAsts)
+import           Development.IDE.GHC.Compat.Util
+import           Ide.Plugin.SelectionRange.ASTPreProcess (PreProcessEnv (PreProcessEnv),
+                                                          preProcessAST)
+import           Ide.Types                               (PluginDescriptor (pluginHandlers),
+                                                          PluginId,
+                                                          defaultPluginDescriptor,
+                                                          mkPluginHandler)
+import           Language.LSP.Server                     (LspM)
+import           Language.LSP.Types                      (List (List),
+                                                          NormalizedFilePath,
+                                                          Position,
+                                                          ResponseError,
+                                                          SMethod (STextDocumentSelectionRange),
+                                                          SelectionRange (..),
+                                                          SelectionRangeParams (..),
+                                                          TextDocumentIdentifier (TextDocumentIdentifier))
+import           Prelude                                 hiding (span)
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId)
+    { pluginHandlers = mkPluginHandler STextDocumentSelectionRange selectionRangeHandler
+    }
+
+selectionRangeHandler :: IdeState -> PluginId -> SelectionRangeParams -> LspM c (Either ResponseError (List SelectionRange))
+selectionRangeHandler ide _ SelectionRangeParams{..} = do
+    let (TextDocumentIdentifier uri) = _textDocument
+    let filePathMaybe = toNormalizedFilePath' <$> uriToFilePath' uri
+    case filePathMaybe of
+        Nothing -> pure . Right . List $ []
+        Just filePath -> liftIO $ do
+            let (List positions) = _positions
+            selectionRanges <- runIdeAction "SelectionRange" (shakeExtras ide) $ getSelectionRanges filePath positions
+            pure . Right . List $ selectionRanges
+
+getSelectionRanges :: NormalizedFilePath -> [Position] -> IdeAction [SelectionRange]
+getSelectionRanges file positions = fmap (fromMaybe []) <$> runMaybeT $ do
```

Comment:
Should we fail the request in some of the cases where we get `Nothing`? Clearly sometimes there are no appropriate selection ranges, but sometimes it might rather be that something went wrong trying to figure them out.

Reply:
The error handling should be done better. I will reconsider that.